[Go Bootcamp] First deliverable

jvillarreal-w opened the pull request:
Had some repetition concerns towards error handling, other than that I just followed the steps of the clean architecture article and added some complementary info to the CSV, such as base stats, so I can fetch that same data from PokeAPI for the next deliverable and not have it be just names.

jvillarreal-w committed:
Initial setup for clean arch  first deliverable.

jvillarreal-w committed:
Made a directory adjustment

jvillarreal-w committed:
Added gitignore

jvillarreal-w committed:
Delete .DS_Store

jvillarreal-w committed:
Changes:

- Removed presenter.
- Added GetById functionality.
- Added registry.
- Added main method logic.

jvillarreal-w committed:
Changes:

- Revamped CSV file initial info.
- Added constants file.
- Adaped repository methods to new CSV format.

jvillarreal-w committed:
Reworked stats into their own struct.

jvillarreal-w committed:
Added small comment.

StradivariT commented on common/constants.go:
```diff
@@ -0,0 +1,5 @@
+package common
+
+const (
+	InternalDataSourcePath = "pkmn.csv"
```
Keeping this kind of info separated is fine, although it's a lot more commonly found to have it on a config file, and then when the app initializes you would like to read the file with this and potentially more values, like the external api endpoints, db tables, etc...

StradivariT commented on infrastructure/router/router.go:
```diff
@@ -0,0 +1,17 @@
+	e.Use(middleware.Logger())
+	e.Use(middleware.Recover())
+
+	e.GET("/pokemon", func(context echo.Context) error { return c.Pokemon.GetPokemon(context) })
```
I'm not familiar with echo, but would it be possible to just have, for example: 
`e.GET("/pokemon", c.Pokemon.GetPokemon(context))` ?

Just to avoid redundancy

StradivariT replied:
Same for the route below

jvillarreal-w replied:
Seems like the intended way of declaring routes with Echo, though to be fair I cannot say I am familiar with this either.
https://echo.labstack.com/guide/routing/

StradivariT commented on interface/controller/app_controller.go:
```diff
@@ -0,0 +1,5 @@
+package controller
+
+type AppController struct {
+	Pokemon interface{ PokemonController }
```
Are you planning on adding more interfaces to the Pokemon attribute? If not you could just remove the `interface` wrapper

jvillarreal-w replied:
Addressed in a more recent commit, thank you.

StradivariT commented on usecase/interactor/pkmn_interactor.go:
```diff
@@ -0,0 +1,35 @@
+}
+
+func (ps *pokemonInteractor) Get(p []*model.Pokemon) ([]*model.Pokemon, error) {
+	p, err := ps.PokemonRepository.FindAll(p)
```
If you don't need to do any more logic after this call then it's ok to just `return ps.PokemonRepository.FindAll(p)`.

Same for the function below

jvillarreal-w replied:
Addressed in a more recent commit, thank you.

StradivariT commented on interface/repository/pkmn_repository.go:
```diff
@@ -0,0 +1,217 @@
+
+	rows, err := csv.NewReader(f).ReadAll()
+	if err != nil {
+		fmt.Println(err)
```
I think you would also like to return here

jvillarreal-w replied:
Addressed as well, thank you.

StradivariT commented on interface/controller/pkmn_controller.go:
```diff
@@ -0,0 +1,49 @@
+package controller
```
This comment applies for all the errors in general.

While it's ok to return when an error happens, what it's usually done is to log the error you got and then respond to the user with an error that would be more appropriate.

Logging allows us to debug from our side what's going on since logs are usually aggregated into a logging service, and it also allows us to not return an error so explicit to the user that could potentially give out details on how our APIs work. So after logging whatever error we got, we want to return to the user a custom error.

For example, if you didn't find the resource the user was requesting it would make sense to return a 404 - not found with a message that goes like `pokemon not found`, or when something internal failed, like failing to read the csv file, or failing to connect with a DB, then we would just send the user a 500 status code with a message like `something went wrong!`, while logging as much info as we can to help us debug if necessary.

So TL;DR, log your errors and return something more user friendly as a response

jvillarreal-w replied:
Added a general purpose logger to address this.

StradivariT commented on interface/repository/pkmn_repository.go:
```diff
@@ -0,0 +1,217 @@
+	"github.com/jvillarreal-w/academy-go-q42021/usecase/repository"
+)
+
+type pokemonRepository struct {
```
I would highly suggest having the file path as an attribute here on your struct, that way when initializing a new Repository, you could provide any file path, which will make unit testing a lot easier since your not binding all of this logic to just one file. We don't want to make tests with the actual real file, but with test files, testing will come later but I thought it would be worth mentioning this now

jvillarreal-w replied:
Added in a later commit, thank you.

StradivariT commented on interface/controller/pkmn_controller.go:
```diff
@@ -0,0 +1,49 @@
+
+func (pc *pokemonController) GetPokemonById(c Context) error {
+	var p *model.Pokemon
+	id := c.Param("id")
```
One of the main responsibilites of a controller is to validate if the data you receive from the user is valid before moving on, if you need to convert this to an int later on, the place to do it is here on your controller, and not in deeper layers of your app. So if the id is not valid, you could return an error response to the user indicating that the id provided is not valid

jvillarreal-w replied:
Now validated in the controller.

StradivariT commented on interface/repository/pkmn_repository.go:
```diff
@@ -0,0 +1,217 @@
+		return nil, err
+	}
+
+	for _, row := range rows {
```
Seems like there's a whole bunch of repeated logic in here, could you try to move the repeated stuff in a reusable function? 

One suggestion I would make is that, if you already have a function that retrieves all of the pokemons, then in this function that finds one by its ID you could call that function and then just search for the requested ID here.

jvillarreal-w replied:
Thanks for the suggestion, that seems to do it.

StradivariT reviewed (changes requested):
Overall looks great, it also seems like it's on a good track to continue with the 2nd deliverable stuff with little changes.

There's stil a couple of things to be done to be able to do unit testing on an ideal way, but so far excellent job!

jvillarreal-w committed:
CHanges:

- Removed Pokemon interface wrapper.
- Simplified interactor calls.
- Added return statement to CSV error handling.

jvillarreal-w committed:
Added a general purpose logger.

jvillarreal-w committed:
Added file path to PokemonRepository

jvillarreal-w committed:
Changes:

- Checking ID validity in controller.
- Addressed repetition in repository.
- Changes for the GetByID function to accept a slice of Pokemon instead.
